### ISSUE TITLE
Use virtual text to show evaluation results

### ISSUE THREAD

**Issue by leana8959** [mrcjkb/haskell-tools.nvim#258]
### Feature description

Hello,

I noticed that with Scala's `metals`, the results of an evaluation show up as virtual text; on the other hand, `haskell-tools` puts them in comments.

Is it possible to have the option to use virtual text instead of comments for evaluation results? I simply think it's neater to not have the file changed when one's simply "probing variables and functions"

A possible problem I can foresee is that longer messages won't fit. A solution could be to place a code action over the evaluated line, which when triggered, shows the complete result.

Here's an image how it looks like for Scala. Note that I can't actually change the text on the right after `//` because it's virtual text.
<img width="554" alt="image" src="https://github.com/mrcjkb/haskell-tools.nvim/assets/87855546/5781b7bc-3740-442b-aebf-7812ae387423">

Thanks, have a nice day :)

**Comment by mrcjkb** [comment]
Hey 😊👋

Thanks for the request.
Unfortunately, I don't think this would be easy to implement in haskell-tools.nvim, as it is [haskell-language-server](https://github.com/haskell/haskell-language-server) that evaluates the code snippet and adds the result to the document.

I like the idea though. Maybe it could be added to haskell-language-server, in the form of a code action that produces diagnostics or something similar that can be displayed as virtual text.
I fear trying to evaluate code snippets on the fly would be too risky (e.g. what happens if a snippet recurses infinitely?)

I'm not sure, but I think what you're seeing in Scala are inlay hints, not actual code snippet evaluations?

**Comment by leana8959** [comment]
> Unfortunately, I don't think this would be easy to implement in haskell-tools.nvim, as it is [haskell-language-server](https://github.com/haskell/haskell-language-server) that evaluates the code snippet and adds the result to the document.

I see, maybe I'll create a feature suggestion in their repo :)

> I fear trying to evaluate code snippets on the fly would be too risky (e.g. what happens if a snippet recurses infinitely?)

That's also a valid concern that I haven't thought of. In `metals` it just shows an error diagnostics, and stop evaluating the rest of the code. I think we might need a way to restart the repl if it loops too long.
<img width="570" alt="image" src="https://github.com/mrcjkb/haskell-tools.nvim/assets/87855546/81350e1d-e5b9-4d53-b17a-8b786642d036">

> I'm not sure, but I think what you're seeing in Scala are inlay hints, not actual code snippet evaluations?

These are "worksheets", I'm not sure if there are equivalent of it in Haskell.
Basically, it evaluates the value of each line and shows hints on the right side.
It's a lot more than just type hints.
<img width="861" alt="image" src="https://github.com/mrcjkb/haskell-tools.nvim/assets/87855546/77992cc6-e460-46f3-8208-6ffd2ee38292">

**Comment by mrcjkb** [comment]
> maybe I'll create a feature suggestion in their repo :)

It definitely sounds like a good idea :smile: 

> In metals it just shows an error diagnostics

Haskell doesn't have a recursion limit (it [doesn't have a traditional call stack](https://wiki.haskell.org/Stack_overflow). So the evaluation would probably have to be an async action with a timeout.

> Basically, it evaluates the value of each line and shows hints on the right side.
It's a lot more than just type hints.

I could imagine this being very practical for pure evaluations (not in `IO`).

> Also thank you for the haskell-tools project, and for taking your time reading my suggestion :)

No problem. I'm happy to share :smile:

**Comment by mrcjkb** [comment]
> These are "worksheets", I'm not sure if there is an equivalent of it in Haskell.

Just read through the [Scala documentation on worksheets](https://docs.scala-lang.org/scala3/book/tools-worksheets.html)

Now I understand... So if I understand correctly, this is available in files with a special extension, but not regular scala files. Hmm, that could be an interesting feature...

Let's leave this issue open and I'll think about whether it makes sense to contribute it to haskell-language-server. But please feel free to open an issue there. Maybe someone will be faster to implement it than me 😊
